### PR TITLE
feat(column)!: rework 'statuscolumn' %r/l items

### LINKED ISSUE #21788: No way to emulate `signcolumn=number` with `statuscolumn`?
### Problem

I like the new `statuscolumn` capability, and am trying to use it to produce a more cleanly-formatted status column.

Prior the option's availablility, I used:

* `set number`
* `set relativenumber`
* `set signcolumn=number`

This worked well, as lines would show a number *unless* there was a sign, in which case that sign would be overlaid in the number column instead. This saves horizontal space, and most numbers are still present (OK if you are just using them for navigation hints).

Unfortunately, as best I can tell, there doesn't seem to be a way of replicating that behaviour with `statuscolumn`. The sign column `%s` format item allows for the sign to be shown, but there's no way to *test* whether a sign is present and conditionally return/show a line number. Possibly I could iterate over all signs in the document (not sure how to get them) but that seems horribly inefficient for something evaluated on every line/redraw.

### Expected behavior

Ideally, I'd be able to write code something like this:

```lua
    _G.custom_statuscol = function()
        if vim.v.has_signs
            return "%s"
        else
            return vim.v.relnum
        end
    end

    vim.opt.statuscolumn = "%{%v:lua.custom_statuscol()%}"                                                                                                                    
    vim.opt.relativenumber = true
    vim.opt.number = true
```

The new element in the code above that doesn't exist in the current featureset (as far as I know) is `vim.v.has_signs`.

Thanks for all the hard work on NeoVim!

### PR DESCRIPTION
**Pull request by luukvbaal**
**refactor: remove redundant copy in statuscolumn items**

Immediately write to buf_tmp instead of to another temporary buffer that
is later copied to buf_tmp. Misc cleanup.

**feat(column)!: rework 'statuscolumn' %r/l items**

Problem:  A custom 'statuscolumn' needs to check a bunch of options and
          placed signs to replicate the default number column.
Solution: Rework %l item to include the necessary logic to mimic the
          default number column. Remove now redundant %r item.

Fix #21788

### CODE REVIEW

**Comment by zeertzjq** [comment]
Can you add some description to the commit message of the first commit?

File: test/functional/ui/statuscolumn_spec.lua [hunk 1]
```diff
@@ -94,46 +94,80 @@ describe('statuscolumn', function()
...
+        {8: 8 }^aaaaa                                             |
+        {8: 9 }aaaaa                                             |
+        {8:10 }aaaaa                                             |
+        {8:11 }aaaaa                                             |
+        {8:12 }aaaaa                                             |
+        {8:13 }aaaaa                                             |
+        {8:14 }aaaaa                                             |
+        {8:15 }aaaaa                                             |
+        {8:16 }aaaaa                                             |
+                                                             |
```
Comment (zeertzjq) [inline test/functional/ui/statuscolumn_spec.lua:111]:
```suggestion
      {8: 4 }aaaaa                                             |
      {8: 5 }aaaaa                                             |
      {8: 6 }aaaaa                                             |
      {8: 7 }aaaaa                                             |
      {8: 8 }^aaaaa                                             |
      {8: 9 }aaaaa                                             |
      {8:10 }aaaaa                                             |
      {8:11 }aaaaa                                             |
      {8:12 }aaaaa                                             |
      {8:13 }aaaaa                                             |
      {8:14 }aaaaa                                             |
      {8:15 }aaaaa                                             |
      {8:16 }aaaaa                                             |
                                                           |
```

File: test/functional/ui/statuscolumn_spec.lua [hunk 2]
```diff
@@ -94,46 +94,80 @@ describe('statuscolumn', function()
...
+        {8:8  }^aaaaa                                             |
+        {8: 1 }aaaaa                                             |
+        {8: 2 }aaaaa                                             |
+        {8: 3 }aaaaa                                             |
+        {8: 4 }aaaaa                                             |
+        {8: 5 }aaaaa                                             |
+        {8: 6 }aaaaa                                             |
+        {8: 7 }aaaaa                                             |
+        {8: 8 }aaaaa                                             |
+                                                             |
```
Comment (zeertzjq) [inline test/functional/ui/statuscolumn_spec.lua:130]:
```suggestion
      {8: 4 }aaaaa                                             |
      {8: 3 }aaaaa                                             |
      {8: 2 }aaaaa                                             |
      {8: 1 }aaaaa                                             |
      {8:8  }^aaaaa                                             |
      {8: 1 }aaaaa                                             |
      {8: 2 }aaaaa                                             |
      {8: 3 }aaaaa                                             |
      {8: 4 }aaaaa                                             |
      {8: 5 }aaaaa                                             |
      {8: 6 }aaaaa                                             |
      {8: 7 }aaaaa                                             |
      {8: 8 }aaaaa                                             |
                                                           |
```

File: test/functional/ui/statuscolumn_spec.lua [hunk 3]
```diff
@@ -94,46 +94,80 @@ describe('statuscolumn', function()
...
+        {8: 0 }^aaaaa                                             |
+        {8: 1 }aaaaa                                             |
+        {8: 2 }aaaaa                                             |
+        {8: 3 }aaaaa                                             |
+        {8: 4 }aaaaa                                             |
+        {8: 5 }aaaaa                                             |
+        {8: 6 }aaaaa                                             |
+        {8: 7 }aaaaa                                             |
+        {8: 8 }aaaaa                                             |
+                                                             |
```
Comment (zeertzjq) [inline test/functional/ui/statuscolumn_spec.lua:149]:
```suggestion
      {8: 4 }aaaaa                                             |
      {8: 3 }aaaaa                                             |
      {8: 2 }aaaaa                                             |
      {8: 1 }aaaaa                                             |
      {8: 0 }^aaaaa                                             |
      {8: 1 }aaaaa                                             |
      {8: 2 }aaaaa                                             |
      {8: 3 }aaaaa                                             |
      {8: 4 }aaaaa                                             |
      {8: 5 }aaaaa                                             |
      {8: 6 }aaaaa                                             |
      {8: 7 }aaaaa                                             |
      {8: 8 }aaaaa                                             |
                                                           |
```

File: src/nvim/statusline.c [hunk 4]
```diff
@@ -1635,50 +1635,43 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, OptIndex op
...
-        if (!fold) {
-          SignTextAttrs *sattr = virtnum ? NULL : &stcp->sattrs[i];
-          p = "  ";
-          if (sattr && sattr->text[0]) {
-            describe_sign_text(buf, sattr->text);
-            p = buf;
+        stl_items[curitem].start = out_p + signlen;
+        if (fdc == 0) {
+          SignTextAttrs *sattrs = get_vim_var_nr(VV_VIRTNUM) ? NULL : &stcp->sattrs[i];
+          if (sattrs->text[0]) {
```
Comment (zeertzjq) [inline src/nvim/statusline.c]: Why is there no NULL check in the condition below when one branch of this ternary operator is NULL?

File: src/nvim/statusline.c [hunk 5]
```diff
@@ -1661,7 +1663,7 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, OptIndex op
         stl_items[curitem].start = out_p + signlen;
         if (fdc == 0) {
           SignTextAttrs *sattrs = get_vim_var_nr(VV_VIRTNUM) ? NULL : &stcp->sattrs[i];
-          if (sattrs->text[0]) {
+          if (sattrs && sattrs->text[0]) {
```
Comment (zeertzjq) [inline src/nvim/statusline.c:1666]: This change should be moved to the first commit.

File: src/nvim/statusline.c [hunk 6]
```diff
@@ -1234,6 +1233,8 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, OptIndex op
     }
     int minwid = 0;
     int maxwid = 9999;
+    int foldsignitem = -1;
+    bool numsepitem = false;
```
Comment (zeertzjq) [inline src/nvim/statusline.c:1237]: The name of this variable is a bit confusing.

File: src/nvim/drawline.c [hunk 7]
```diff
@@ -587,12 +587,13 @@ static void draw_statuscol(win_T *wp, winlinevars_T *wlv, linenr_T lnum, int vir
...
-  // width of the status column by building with "w_nrwidth_line_count". Add
-  // potentially truncated width and rebuild before drawing anything.
+  // width of the status column by building with the largest possible line number.
+  // Add potentially truncated width and rebuild before drawing anything.
   if (wp->w_statuscol_line_count != wp->w_nrwidth_line_count) {
     wp->w_statuscol_line_count = wp->w_nrwidth_line_count;
     set_vim_var_nr(VV_VIRTNUM, 0);
-    int width = build_statuscol_str(wp, wp->w_nrwidth_line_count, 0, buf, stcp);
+    int maxlnum = MAX(wp->w_nrwidth_line_count, wp->w_height_inner);
+    int width = build_statuscol_str(wp, maxlnum, maxlnum, buf, stcp);
```
Comment (zeertzjq) [inline src/nvim/drawline.c:596]: Why do `lnum` and `relnum` have the same value here? Shouldn't it be like this?
```suggestion
    int width = build_statuscol_str(wp, wp->w_nrwidth_line_count, wp->w_height_inner, buf, stcp);
```

Comment (luukvbaal) [reply src/nvim/drawline.c:596]: I agree that seems logical but it doesn't work for 'relativenumber' + 'number' where the line count will only be substituted for this estimation when `relnum == 0`, like before. Whereas `relnum == 0` does not yield the right width when the window height is larger than the line count.

I think this together with the logic for the %l item always yields the largest possible width that may be used somewhere in the window.

File: test/functional/ui/statuscolumn_spec.lua [hunk 8]
```diff
@@ -912,8 +962,8 @@ describe('statuscolumn', function()
     command('call setline(1, range(1, 99))')
```
Comment (zeertzjq) [inline test/functional/ui/statuscolumn_spec.lua:962]: Maybe this test should use `v:lnum` instead of `%l` now?

Comment (luukvbaal) [reply test/functional/ui/statuscolumn_spec.lua:962]: Irrelevant I think but sure.

**Pull request merged**